### PR TITLE
[hail] unify build systems a bit (and fix references.html)

### PR DESCRIPTION
**Pull request by danking**
Our build system is a bit of a tangled mess. There's `build.yaml`, `build.gradle`, `hail/Makefile`, and `src/main/c/Makefile`. In this PR, I've tried to minimize `build.gradle`'s role to just JVM compilation and JAR production. I've removed a number of shell scripts that were previously gluing together build steps. I've either removed the need for that glue or incorporated it into `hail/Makefile`. I did not attempt to unify `src/main/c/Makefile` with `hail/Makefile`.

I started to unify `build.yaml` and `hail/Makefile` but it grew into a bigger project than I want to tackle.

A couple things I expect to raise questions:

Q. Where did COMPILE_NATIVES go?
A: I moved tightly around native-lib, which is the only rule it affects

Q: Why did you remove env/SHORT_REVISION from $(SHADOW_JAR)'s dependencies?
A: It is not an immediate dependency, it's transitive through the $(SCALA_BUILD_INFO)

Q: Why did you remove `define properties ...`?
A: AFAIK, this had no effect and was unused

Q: Why inline $(DATE)?
A: Obviated the need for a comment explaining why we didn't make it an `env_var`.

Q: Why no parallelism in pytest and doctest?
A: Hail uses all cores by default. Getting parallelism right in pytest and doctest would require a bit of smarts. I'm doubt its worth the effort since we have CI and you can test specific tests.

Q: What the hell is `pgradle` and the `+` prefix?
A: This uses the [Make job server](https://www.gnu.org/software/make/manual/html_node/POSIX-Jobserver.html#POSIX-Jobserver) to cooperatively obey parallelism limits (`--jobs`). It uses a pair of pipes to claim and return *extra* CPU reservations (we implicitly have one core). Hail's JVM tests use two core clusters, so I set PARALLELISM to half the number of cores we could claim (with a minimum of one). I did not use this for shadowJar because there is no parallelism to be had.

Q: How did you fix references.html?
A: `build.yaml` and `hail/Makefile` both use `www/Makefile` which now correctly finds all the markdown files instead of a limited list.

Q: Why did you change conftest.py?
A: It is now location independent, so you can run doctests using `python3 setup.py pytest ...`

### CODE REVIEW

**Comment by danking** [comment]
cc: @gsarma

**Comment by jigold** [comment]
Assigning to @chrisvittal because he already looked through most of this and has opinions about it.

File: hail/Makefile [hunk 1]
```diff
@@ -207,8 +233,59 @@ deploy: tag
...
+.PHONY: docs
+docs: www install $(PYTHON_VERSION_INFO)
+	mkdir -p build
+	cp www build/www
+	mkdir -p build/docs
+	cp -R python/hail/docs build/docs
+	sed -E "s/\(hail\#([0-9]+)\)/(\[#\1](https:\/\/github.com\/hail-is\/hail\/pull\/\1))/g" \
+    < build/docs/change_log.md \
+    | pandoc -o build/docs/change_log.rst
+	+$(MAKE) -C www
```
Comment (chrisvittal) [inline hail/Makefile]: I don't know about make 3.81, but `+` is superfluous for `$(MAKE)` invocations for my version.

**Comment by danking** [comment]
Grr. I realized that our build.yaml `build_hail` step did not explicitly depend on setting up the python package correctly, it just relied on `make jars` doing that as a side effect. I've added an explicit rule for setting up the version files and I call it in build.yaml.

The better answer is to change build.yaml to ship around a wheel file. I don't want to do this right now.

File: hail/Makefile [hunk 2]
```diff
@@ -61,6 +61,8 @@ src/main/resources/build-info.properties: Makefile
 	echo 'sparkVersion=$(SPARK_VERSION)' >> $@
 	echo 'hailPipVersion=$(HAIL_PIP_VERSION)' >> $@
 
+python-version-info: $(PYTHON_VERSION_INFO)
```
Comment (chrisvittal) [inline hail/Makefile]: .PHONY: ?

**Comment by chrisvittal** [comment]
#6927 Should fix the issue I was seeing with the warnings. The issue was that to the `make` invocations that we were making in the compilation of C++ pipelines, it looked like they were in jobserver mode. But they we didn't preserve the right state to execute them. 

I think the real solution is unsetting `MAKEFLAGS` in the environment in `pgradle`.

**Comment by danking** [comment]
> This is because our C++ backend uses make to drive compilation (wtf‽).
> (wtf?).
Yeah. My reaction too.

File: hail/Makefile (hunk 2 above)
Comment (danking) [reply hail/Makefile]: done

File: hail/Makefile (hunk 1 above)
Comment (danking) [reply hail/Makefile]: removed

File: hail/pgradle [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+import sys
+
+MAX_PARALLELISM = 16
+make_flags = os.environ.get('MAKEFLAGS')
+
+write_fd = None
+job_slots = b''
+try:
+    if make_flags is not None:
+        match = re.search(' --jobserver-fds=([^,]+),([^ ])+', make_flags)
```
[the remark this answers is not in the document]
Comment (danking) [reply hail/pgradle]: goddammit apple. my default make uses fds

**Comment by danking** [comment]
> Your tool should also examine the first word of the MAKEFLAGS variable and look for the character n. If this character is present then make was invoked with the ‘-n’ option and your tool should stop without performing any operations.

Added.

> Your tool should be sure to write back the tokens it read, even under error conditions. This includes not only errors in your tool but also outside influences such as interrupts (SIGINT), etc. You may want to install signal handlers to manage this write-back.

I mean, I doubt anyone is sending signals other than SIGKILL to our build system, but I added some signal handlers that just `sys.exit(0)` which triggers the finally (I checked).

> We also get a lot of ‘warning: jobserver unavailable: using -j1. Add +' to parent make rule.’warnings when runningmake jvm-test`. This is because our C++ backend uses make to drive compilation (wtf‽).

I strip MAKEFLAGS before calling gradle now.

**Pull request merged**